storage-operator/pkg/apis/storage/v1alpha1/volume_types.go

```diff
@@ -42,12 +48,88 @@ type VolumeSpec struct {
 	VolumeSource `json:",inline"`
 }
 
+type PreparationStep string
+
+// "Enum" representing the preparation steps of a volume.
+const (
+	PreparationStarted PreparationStep = "PreparationStarted"
+	PrepareVolume      PreparationStep = "PrepareVolume"
+	GetVolumeSize      PreparationStep = "GetVolumeSize"
+	CreatePV           PreparationStep = "CreatePersistentVolume"
+)
+
+type FinalizationStep string
+
+// "Enum" representing the preparation steps of a volume.
+const (
+	FinalizationStarted FinalizationStep = "FinalizationStarted"
+	UnprepareVolume     FinalizationStep = "UnprepareVolume"
+	DeletePV            FinalizationStep = "DeletePersistentVolume"
+)
+
+type VolumePhase string
+
+// "Enum" representing the preparation steps of a volume.
+const (
+	VolumeAvailable   VolumePhase = "Available"
+	VolumePending     VolumePhase = "Pending"
+	VolumeFailed      VolumePhase = "Failed"
+	VolumeTerminating VolumePhase = "Terminating"
+)
+
+type VolumeErrorCode int
+
+// "Enum" representing the error codes of the Failed state.
+const (
+	InternalError    VolumeErrorCode = 0
+	CreationError    VolumeErrorCode = 1
+	DestructionError VolumeErrorCode = 2
+	UnavailableError VolumeErrorCode = 3
+)
+
+// Context associated to the Pending phase.
+type VolumeStatusContextPending struct {
+	// Name of the current step.
+	// +kubebuilder:validation:Enum=PreparationStarted,PrepareVolume,GetVolumeSize,CreatePersistentVolume
+	Step PreparationStep `json:"step,omitempty"`
+
+	// JID of the Salt job in progress.
+	JobId string `json:"jobId,omitempty"`
+
+	// Size of the volume.
+	Size resource.Quantity `json:"size,omitempty"`
+}
+
+// Context associated to the Failed phase.
+type VolumeStatusContextFailed struct {
+	// +kubebuilder:validation:Enum=0,1,2,3
+	ErrorCode VolumeErrorCode `json:"errorCode,omitempty"`
+	ErrorMsg  string          `json:"errorMessage,omitempty"`
+}
+
+// Context associated to the Terminating phase.
+type VolumeStatusContextTerminating struct {
+	// Name of the current step.
+	// +kubebuilder:validation:Enum=FinalizationStarted,UnprepareVolume,DeletePersistentVolume
```

Comment:
Too bad these strings need to be copied into this comment ;)

Reply:
Yup, given that Go doesn't have a real enum type I guess that's the only way :/